tap_spreadsheets_anywhere/s3_batch_messages.py

```diff
@@ -0,0 +1,140 @@
+    sys.stdout.flush()
+
+
+def _copy_file(s3, src_bucket: str, dest_bucket: str, key: str, new_key: str, multipart_threshold=5*1024**3):
+    """
+    AWS S3 threshold for a single object copy is 5GB therefore the default value for multipart_threshold is also 5GB.
+    """
+    copy_source = {
+        'Bucket': src_bucket,
+        'Key': key
+    }
+    
+    try:
+        metadata = s3.head_object(Bucket=src_bucket, Key=key)
+
+        # If the source object size is greater than or equal to multipart_threshold,
+        # use multipart copy
+        if metadata['ContentLength'] >= multipart_threshold:
+            multipart_upload = s3.create_multipart_upload(Bucket=dest_bucket, Key=new_key)
+
+            position = 0
+            part_number = 1
+            parts = []
+
+            while position < metadata['ContentLength']:
+                copy_range = f'bytes={position}-{min(position + multipart_threshold - 1, metadata["ContentLength"] - 1)}'
+
+                copy_part = s3.upload_part_copy(
+                    Bucket=dest_bucket,
+                    Key=new_key,
+                    PartNumber=part_number,
+                    UploadId=multipart_upload['UploadId'],
+                    CopySource=copy_source,
+                    CopySourceRange=copy_range,
+                )
+
+                parts.append({
+                    'PartNumber': part_number,
+                    'ETag': copy_part['CopyPartResult']['ETag']
+                })
+
+                position += multipart_threshold
+                part_number += 1
+
+            s3.complete_multipart_upload(
+                Bucket=dest_bucket,
+                Key=new_key,
+                UploadId=multipart_upload['UploadId'],
+                MultipartUpload={'Parts': parts}
+            )
+        else:
+            s3.copy(copy_source, dest_bucket, new_key)
+    except BotoCoreError as e:
+        raise Exception(f'Error occured when copying {src_bucket}/{key} object to stage bucket:\n{e}')
+
+
+def copy_files_to_stage(config: Dict, target_files: List[Dict], source_bucket: str, destination_bucket: str) -> None:
+    LOGGER.info(f'Copying {len(target_files)} files from {source_bucket} to {destination_bucket} stage bucket...')
+    
+    s3_arn_role = config.get('s3_arn_role')
+    if s3_arn_role:
+        client = boto3.client('sts')
+        credentials = client.assume_role(RoleArn=s3_arn_role,
+                                         RoleSessionName='SourceToStageELT')['Credentials']
+
+        s3_client = boto3.resource(
+            's3',
+            aws_access_key_id=credentials['AccessKeyId'],
+            aws_secret_access_key=credentials['SecretAccessKey'],
+            aws_session_token=credentials['SessionToken']
+        ).meta.client
+    else:
+        s3_client = boto3.client('s3')
+    
+    for tfile in target_files:
+        key = tfile['key']
+        _copy_file(s3_client, source_bucket, destination_bucket, key, key)
```

Comment:
*question:* it will probably make a sense to put all copied files under some folder, should it be a `source_bucket` name or something else?

Reply:
I have made a new key to have a top-level folder to be `source_bucket` name.